doc/link_event_damping/Link-event-damping-HLD.md

```diff
@@ -0,0 +1,480 @@
+    - At t=105: link-down event, the current accumulated penalty is decayed to 1088 and it increases to 2088 and the new accumulated penalty of 2088 goes above the suppress threshold (=1600), so the damping starts again on the interface. Link-down event is advertised to applications and the operational state is set as DOWN.
+    - Subsequent link state events are not advertised as accumulated penalty remains above reuse threshold.
+    - The operational state stays DOWN after t = 105.
+    - At t = 115, the accumulated penalty increment to a maximum ceiling of 4800.
+- t = 124: final link event of link-up received
+    - The current accumulated penalty is 4376 after decay
+    - Since damping is active on the interface, the link-up event is suppressed. The operational state stays DOWN.
+- t = 152: accumulated penalty decays below the reuse threshold (=1200)
+    - The accumulated penalty decays below the reuse threshold(=1200), so the damping on the interface stops.
+    - Since the interface is no longer damped and the current damped link-event is UP, the link-up event is advertised to applications. The operational state is set as UP.
+
+
+### SAI API
+
+Feature requires the host interface’s operational state update to be consistent with post damping behavior of the physical interface’s link event. In this proposal we are expecting the host interface’s operational state to be updated by OA application using hostif SAI API when it receives the post damping advertisement of physical interface link event.
+
+Following SAI API changes are proposed to program the modes to update the host interface operational state:
+
+**File: saiswitch.h**
+```
+typedef enum _sai_switch_attr_t
+{
+.
+.
+    /**
+     * @brief Set hostif operational status update mode.
+     *
+     * This will set layer responsible for updating the operational status of
+     * hostif.
+     *
+     * @type sai_switch_hostif_oper_status_update_mode_t
+     * @flags CREATE_AND_SET
+     * @default SAI_SWITCH_HOSTIF_OPER_STATUS_UPDATE_MODE_APPLICATION
+     */
+    SAI_SWITCH_ATTR_HOSTIF_OPER_STATUS_UPDATE_MODE,
+.
+.
+} sai_switch_attr_t;
+
+```
+where `sai_switch_hostif_oper_status_update_mode_t` is defined as:
+
+```
+/**
+ * @brief Attribute data for SAI_SWITCH_ATTR_HOSTIF_OPER_STATUS_UPDATE_MODE.
+ */
+typedef enum _sai_switch_hostif_oper_status_update_mode_t
+{
+    /**
+     * In this mode, operational status of hostif must be updated by application
+     * using hostif API with SAI_HOSTIF_ATTR_OPER_STATUS attribute. SAI adapter
+     * should not update the hostif operational status internally.
+     */
+
+
+    SAI_SWITCH_HOSTIF_OPER_STATUS_UPDATE_MODE_APPLICATION = 0,
+    /**
+     * In this mode, operational status of hostif is updated internally by SAI
+     * adapter. Update of hostif operational status by application using hostif
+     * API with SAI_HOSTIF_ATTR_OPER_STATUS is ignored.
+     */
+    SAI_SWITCH_HOSTIF_OPER_STATUS_UPDATE_MODE_SAI_ADAPTER = 1,
```

Comment:
@Ashish1805 is this for different implementation of damping across vendor? I think we want one implementation logic/algorithm for damping.

Reply:
As Eddy commented, this option is to provide flexibility in choosing syncd vs SAI vendor layer implementation. When mode SAI_SWITCH_HOSTIF_OPER_STATUS_UPDATE_MODE_SAI_ADAPTER is used, SAI implementation is used. Link event damping config will have damping algorithm as an input so depending on the algo in config, SAI vendor layer will use that algorithm for damping.